src/DataStructures/Tensor/EagerMath/DotProduct.hpp

```diff
@@ -0,0 +1,65 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines functions euclidean dot_product and dot_product with a metric
+
+#pragma once
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+/*!
+ * \ingroup TensorGroup
+ * \brief compute the Euclidean dot product of two rank-1 tensors
+ *
+ * \returns the Euclidean dot product of the rank-1 tensors
+ *
+ * \details
+ * Returns \f$A^a B^b \delta_{ab}\f$ for input tensors \f$A^a\f$ and \f$B^b\f$.
+ */
+template <typename DataType, typename Index>
+DataType dot_product(
```

Comment:
I think we want the return type to be a `Scalar<DataType>` as opposed to just a `DataType`. I'd say it communicates intent better (since this is a physical quantity) and I also think that it will play better with other parts of the code (I seem to recall Variables can only be constructed with Tensors, and we had an issue on the bitbucket version of the code that only Tensors could be observed).

Reply:
magnitude() returns DataType, so for consistency I figured dot_product() should as well. Are you saying you want magnitude() to return DataType but DotProduct to return Scalar<DataType>?

Reply:
I think magnitude should return DataType and a contraction should return a Scalar<DataType>.

Reply:
I’m fine with returning a scalar if that’s the consensus. Magnitude and dot product both are contractions with a metric. The only difference is magnitude puts one input vector into both slots of the metric, while for product puts one vector into each slot.

Reply:
Well if someone asked me for the "magnitude" of a Tensor wouldn't know what to give them, whereas a contraction is clearer, that's my reasoning for wanting them to return different things. `magnitude` will likely survive past TensorExpressions in the sense of doing different types of norms/magnitudes of Tensors, a contraction function won't.

Reply:
Stupid autocorrect. That should say dot product puts one each of two input vectors into each slot of the metric, while magnitude puts the only input vector into both slots.

Reply:
If magnitude() becomes a more general norm() function, then sure, it’s different. But as coded, and as the word magnitude is commonly used, the meaning is unambiguous. Magnitude of a tensor and a metric is unambiguous: it’s the square root of the number you get when putting the vector into both of the metric’s slots. Currently the metric just defaults to kronecker delta if none is specified to magnitude(). I’d advocate for a TensorNorn() class to handle L2, LInf, etc. norms.

Reply:
So you're saying we should delete `magnitude`? I disagree, if someone asks for the magnitude of a Tensor I wouldn't know what to give them. That's fine by me and actually probably better than having both a magnitude and a dot product that do the same thing. Also, why not just do a general contraction instead of a dot product for only vectors?

Reply:
Similarly, dot product of two vectors given a metric is unambiguous: put each vector into one slot of the metric.

Reply:
I’m not advocating deleting magnitude, because it is the most common dot product we use, making a direct call to it convenient. But its implementation could call dot_product() if you insist. The code is so straightforward and unlikely to change not sure it is worth the time to make them share one implementation. 

Dot products come up fairly often in GR calculations. Right now, I need a combination of 3 dot products to compute the area element on a strahlkorper.

Reply:
I have 2 requests:

1) Remove magnitude, it's redundant and the extra argument won't kill anyone to type
2) generalize this to be able to contract arbitrary indices of the same types

Reply:
1. It’s not a matter of an extra argument. You’re proposing we globally replace magnitude(x, g) -> sqrt(dot_product(x,x,g)), which is just too inconvenient for such a commonly used thing.

2. It’s better not to get hung up on contraction here. I’m talking about a binary operation on two vectors. You can technically think of a dot product as forming a rank 4 tensor and then contracting, but it’s clearer to just understand it as an operation defined by putting vectors into the slots of the metric. Isn’t what you’re really asking for a generic raise/lower tensor index function (which we’ve avoided so far), which would then be called when needed by a generic contract these indices function? That’s way larger in scope than necessary here, and it would make magnitude even more cumbersome to code, replacing magnitude(x,g) -> contract(contract(tensor_product(tensor_product(x,x),g),0,2),1,3).

Reply:
1. It's not that common, you should be grabbing the result out of the DataBox in most cases anyway.

2. Yes, that is what I'm suggesting. You can wait for others to weigh in.

Reply:
Yes, good idea; let’s discuss on tomorrow’s spectre call.

Reply:
I'd prefer a record of the discussion here.

Reply:
My opinions:
* We should keep magnitude because I want to use it as the `function` for a compute item so removing it will just make me rewrite it inside the tag, and if we have to maintain the code anyway we might as well let people take advantage of it.
* Both dot_product and magnitude should return Scalar.
* dot_product should replace the proposed magnitude_squared from #534.
* Taking a dot product as outer products followed by contractions is horribly inefficient without something like a smart tensor expression library.  I don't know whether the minimal one we have can handle it efficiently (or at all), since it is basically undocumented, but a quick test seems to indicate that it can't.  So we shouldn't do that.

Reply:
I like what @wthrowe proposes. I’m fine with both magnitude and dot_product returning a scalar, but I suggest that @osheamonn do a separate pull request making the change for magnitude, since he wrote the original function.

Reply:
Yes, I don't think the magnitude change needs to be done in this PR.

Reply:
No, it should be done in this PR

Reply:
I agree with @wthrowe. magnitude() is used many places in the code, so making the change requires changing many parts of the code. That should be a separate, narrow pull request that is made, checked and approved rapidly. Otherwise, it becomes a big burden to maintain on a fork as the code continues to evolve.

Reply:
Are you volunteering to do it?

Reply:
I'm working on it now.

Reply:
And if I understand what Will is suggesting, he's suggesting having magnitude call dot_product, which does not require a lot of code changes

Reply:
Yes, magnitude calling dot_product is easy. But changing return to scalar is hard because all uses go from magnitude() to get(magnitude()). Thanks for already getting started on changing magnitude(), @osheamonm!